File: week6-network-flow/maxflow.py
import sys

class edge: 
    def __init__(self, cap, usage, to, direction):
        self.cap: int = cap
        self.usage: int = usage
        self.to: int = to
        self.direction: bool = direction
    def __repr__(self):
        return f'Edge(capacity={self.cap}, usage={self.usage}, to={self.to})'
        
class vertex: 
    def __init__(self, id):
        self.id: int = id
        self.edges: list[edge] = []
        self.sink = False
        self.source = False

    def set_sink(self, sink: bool):
        self.sink = sink

    def set_source(self, source: bool):
        self.source = source
    
    def add_edge(self, edge: edge):
        self.edges.append(edge)

    def __repr__(self):
        edges_repr = "\n"
        edges_repr += "\n ".join(str(edge) for edge in self.edges)
        return f'Vertex(id={self.id}, source={self.source}, sink={self.sink}, edges=[{edges_repr}])'

def findpath(graph: list[vertex], u: int, visited: set[int], path: list[edge], t: int) -> bool:
    """DFS to find an augmenting path from the source to the sink."""
    if u == t:
        return True
    visited.add(u)
    for e in graph[u].edges:
        residual_capacity = e.cap - e.usage
        if residual_capacity > 0 and e.to not in visited:
            path.append(e)
            if findpath(graph, e.to, visited, path, t):
                return True
            path.pop() 
    return False

def ford_fulkerson(graph: list[vertex], source: int, sink: int) -> int:
    """Implementation of Ford-Fulkerson to find maximum flow."""
    max_flow = 0
    path = []
    
    while findpath(graph, source, set(), path, sink):
        flow = min(e.cap - e.usage for e in path)
        for e in path:
            e.usage += flow
            reverse_edge = next((re for re in graph[e.to].edges if re.to == graph[e.to].id and re.usage < 0), None)
            if reverse_edge is None:
                graph[e.to].add_edge(edge(cap=0, usage=-flow, to=e.to, direction=True)) 
            else:
                reverse_edge.usage -= flow
        max_flow += flow
        path.clear()  

    return max_flow, graph

if len(sys.argv) > 1:
    with open(sys.argv[1], 'r') as file:
        input_data = file.read().strip().split('\n')
else:
    input_data = sys.stdin.read().strip().split('\n')

n, m, s, t = map(int, input_data[0].split())
'''
n = vertices
m = edges
s = source
t = sink
'''
graph: list[vertex] = [vertex(i) for i in range(n)]
graph[s].set_source(True)
graph[t].set_sink(True)

for i in range(m):
    u, v, x = map(int, input_data[i + 1].split())
    graph[u].add_edge(edge(x, 0, v, False))
    graph[v].add_edge(edge(0, 0, u, True)) 

max_flow,graph = ford_fulkerson(graph, s, t)
used = 0 

if(max_flow!=0):
    out = ""
    for e in graph:
        for ee in e.edges:
            if(ee.usage != 0 and not ee.direction):
                out += f'{e.id} {ee.to} {ee.usage}\n'
                used +=1
    out = out[:-1]

print(f'{n} {max_flow} {used}')

if(max_flow!=0):
    print(out)

File: week6-network-flow/maxflow.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Edge struct {
	from, to, cap, flow int
}

type Graph struct {
	edges   []Edge
	adj     [][]int
	level   []int
	ptr     []int
	n, source, sink int
}

func NewGraph(n, source, sink int) *Graph {
	return &Graph{
		edges:   make([]Edge, 0),
		adj:     make([][]int, n),
		level:   make([]int, n),
		ptr:     make([]int, n),
		n:       n,
		source:  source,
		sink:    sink,
	}
}

func (g *Graph) AddEdge(from, to, cap int) {
	g.edges = append(g.edges, Edge{from, to, cap, 0})
	g.edges = append(g.edges, Edge{to, from, 0, 0}) // Reverse edge
	g.adj[from] = append(g.adj[from], len(g.edges)-2) // Forward edge index
	g.adj[to] = append(g.adj[to], len(g.edges)-1)     // Reverse edge index
}

func (g *Graph) BFS() bool {
	for i := range g.level {
		g.level[i] = -1
	}
	g.level[g.source] = 0
	queue := []int{g.source}

	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]

		for _, edgeIndex := range g.adj[u] {
			edge := &g.edges[edgeIndex]
			if g.level[edge.to] == -1 && edge.flow < edge.cap {
				g.level[edge.to] = g.level[u] + 1
				queue = append(queue, edge.to)
				if edge.to == g.sink {
					return true
				}
			}
		}
	}
	return false
}

func (g *Graph) DFS(u, flow int) int {
	if u == g.sink {
		return flow
	}
	for g.ptr[u] < len(g.adj[u]) {
		edgeIndex := g.adj[u][g.ptr[u]]
		edge := &g.edges[edgeIndex]
		if g.level[edge.to] == g.level[u]+1 && edge.flow < edge.cap {
			currentFlow := min(flow, edge.cap-edge.flow)
			if pushed := g.DFS(edge.to, currentFlow); pushed > 0 {
				edge.flow += pushed
				g.edges[edgeIndex^1].flow -= pushed // Update reverse edge
				return pushed
			}
		}
		g.ptr[u]++
	}
	return 0
}

func (g *Graph) MaxFlow() (int, int) {
	totalFlow := 0

	for g.BFS() {
		for i := range g.ptr {
			g.ptr[i] = 0
		}
		for {
			flow := g.DFS(g.source, 1<<30) // Some large number
			if flow == 0 {
				break
			}
			totalFlow += flow
		}
	}
	return totalFlow, g.getUsedEdges()
}

func (g *Graph) getUsedEdges() int {
	used := 0
	for _, edge := range g.edges {
		if edge.flow > 0 && edge.from != edge.to {
			used++
		}
	}
	return used
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	var n, m, s, t int
	var inputData []string

	if len(os.Args) > 1 {
		file, err := os.Open(os.Args[1])
		if err != nil {
			fmt.Println(err)
			return
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			inputData = append(inputData, scanner.Text())
		}
	} else {
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			inputData = append(inputData, scanner.Text())
		}
	}

	fmt.Sscanf(inputData[0], "%d %d %d %d", &n, &m, &s, &t)

	graph := NewGraph(n, s, t)

	for i := 1; i <= m; i++ {
		var u, v, x int
		fmt.Sscanf(inputData[i], "%d %d %d", &u, &v, &x)
		graph.AddEdge(u, v, x)
	}

	maxFlow, usedEdges := graph.MaxFlow()
	out := ""

	if maxFlow != 0 {
		for _, edge := range graph.edges {
			if edge.flow > 0 && edge.from != edge.to {
				out += fmt.Sprintf("%d %d %d\n", edge.from, edge.to, edge.flow)
			}
		}
		out = out[:len(out)-1] // Remove the trailing newline

		fmt.Printf("%d %d %d\n", n, maxFlow, usedEdges)
		fmt.Print(out)
	} else {
		fmt.Printf("%d %d %d\n", n, maxFlow, usedEdges)
	}
}
